unit_scaling/transforms/utils.py

```diff
@@ -0,0 +1,135 @@
+# Copyright (c) 2023 Graphcore Ltd. All rights reserved.
+
+import functools
+from contextlib import contextmanager
+from copy import copy
+from typing import Any, Callable, Dict, Iterable, Optional, Set, Tuple, TypeVar
+from unittest.mock import patch
+
+import torch._dynamo
+from torch import nn
+from torch.fx.graph import Graph
+from torch.fx.node import Node
+
+from .._internal_utils import generate__all__
+
+T = TypeVar("T")
+
+
+def _get_patched_allowed_function_ids(
+    non_recurse_functions: Iterable[Callable[..., Any]],
+) -> Set[int]:
+    allowed_function_ids = copy(torch._dynamo.allowed_functions._allowed_function_ids)
+    for v in nn.modules.__dict__.values():
+        if isinstance(v, type) and v not in nn.modules.loss.__dict__.values():
+            i = id(v)
+            if i in allowed_function_ids:
+                allowed_function_ids.remove(i)
+    for f in non_recurse_functions:
+        allowed_function_ids.add(id(f))
+    return allowed_function_ids  # type: ignore[no-any-return]
+
+
+def _patched_call_function(self, tx, args, kwargs):  # type: ignore[no-untyped-def]
+    if tx.output.is_root_tracer() and isinstance(
+        self.obj, torch._dynamo.variables.NNModuleVariable
+    ):
+        module_attr = getattr(self.fn, "__module__", "")
+        if (
+            module_attr is not None
+            and module_attr.startswith("torch.nn.modules.module")
+            or self.is_constant
+        ):
+            return self.obj.call_method(  # type: ignore[no-untyped-call]
+                tx, self.fn.__name__, args, kwargs, constant=self.is_constant
+            ).add_options(self)
+    return super(
+        torch._dynamo.variables.functions.UserMethodVariable, self
+    ).call_function(tx, args, kwargs)
+
+
+@contextmanager
+def _expand_modules_patch(non_recurse_functions):  # type: ignore[no-untyped-def]
+    patcher_a = patch(
+        "torch._dynamo.allowed_functions._allowed_function_ids",
```

Comment:
This is very likely to break in future PyTorch versions, but should tell us when it does!